src/overlays/actors/ovl_Bg_Ice_Shutter/z_bg_ice_shutter.c

```diff
@@ -27,19 +31,105 @@ const ActorInit Bg_Ice_Shutter_InitVars = {
+    thisx->params = (thisx->params >> 8) & 0xFF;
+    DynaPolyInfo_Alloc(&D_06002854, &localC);
+    this->dyna.dynaPolyId = DynaPolyInfo_RegisterActor(globalCtx, &globalCtx->colCtx.dyna, thisx, localC);
+    if (sp28 == 2) {
+        thisx->shape.rot.x = -0x4000;
+    }
+
+    if (sp28 != 1) {
+        if (Flags_GetClear(globalCtx, thisx->room)) {
+            Actor_Kill(thisx);
+        } else {
+            this->actionFunc = func_80891CF4;
+        }
+
+    } else {
+        if (Flags_GetSwitch(globalCtx, thisx->params)) {
+            Actor_Kill(thisx);
+        } else {
+            this->actionFunc = func_80891D6C;
+        }
+    }
+
+    if (sp28 == 2) {
+        temp_f6 = Math_Sins(thisx->shape.rot.x) * 50.0f;
+        thisx->posRot2.pos.x = (Math_Sins(thisx->shape.rot.y) * temp_f6) + thisx->initPosRot.pos.x;
+        thisx->posRot2.pos.y = thisx->initPosRot.pos.y;
+        thisx->posRot2.pos.z = thisx->initPosRot.pos.z + (Math_Coss(thisx->shape.rot.y) * temp_f6);
+    } else {
+        Actor_SetHeight(thisx, 50.0f);
+    }
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Bg_Ice_Shutter/BgIceShutter_Destroy.s")
+void BgIceShutter_Destroy(Actor* thisx, GlobalContext* globalCtx) {
+    DynaPolyInfo_Free(globalCtx, &globalCtx->colCtx.dyna, THIS->dyna.dynaPolyId);
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Bg_Ice_Shutter/func_80891CF4.s")
+void func_80891CF4(BgIceShutter* this, GlobalContext* globalCtx) {
+    if (Flags_GetTempClear(globalCtx, this->dyna.actor.room) != 0) {
+        Flags_SetClear(globalCtx, this->dyna.actor.room);
+        Audio_PlaySoundAtPosition(globalCtx, &(this->dyna.actor.posRot), 0x1E, 0x2814);
+        this->actionFunc = &func_80891DD4;
+        if (this->dyna.actor.shape.rot.x == 0) {
+            func_80080480(globalCtx, this);
+        }
+    }
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Bg_Ice_Shutter/func_80891D6C.s")
+void func_80891D6C(BgIceShutter* this, GlobalContext* globalCtx) {
+    if (Flags_GetSwitch(globalCtx, this->dyna.actor.params) != 0) {
+        Audio_PlaySoundAtPosition(globalCtx, &(this->dyna.actor.posRot), 0x1E, NA_SE_EV_SLIDE_DOOR_OPEN);
+        this->actionFunc = &func_80891DD4;
+        func_80080480(globalCtx, this);
+    }
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Bg_Ice_Shutter/func_80891DD4.s")
+void func_80891DD4(BgIceShutter* this, GlobalContext* globalCtx) {
+    Math_ApproxF(&this->dyna.actor.speedXZ, 30.0f, 2.0f);
+    if (Math_ApproxF(&this->dyna.actor.velocity.y, 210.0f, this->dyna.actor.speedXZ)) {
+        Actor_Kill(this);
+        return;
+    }
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Bg_Ice_Shutter/BgIceShutter_Update.s")
+    func_80891AC0(this);
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Bg_Ice_Shutter/BgIceShutter_Draw.s")
+void BgIceShutter_Update(Actor* thisx, GlobalContext* globalCtx) {
+    THIS->actionFunc(THIS, globalCtx);
```

Comment:
Does the stack allow a temp to add `BgIceShutter* this = THIS;` to use `this` instead?